### PR TITLE
add list_column_families api

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
add list_column_families api

### CODE REVIEW

**Comment by zhangjinpeng1987** [comment]
@siddontang @BusyJay @ngaut PTAL

File: librocksdb_sys/src/lib.rs [hunk 1]
```diff
@@ -440,6 +440,13 @@ extern "C" {
                                       column_family_handle: DBCFHandle,
                                       err: *mut *const i8);
     pub fn rocksdb_column_family_handle_destroy(column_family_handle: DBCFHandle);
+    pub fn rocksdb_list_column_families(db: DBOptions,
+                                        path: *const i8,
+                                        lencf: *const size_t,
```
Comment (BusyJay) [inline librocksdb_sys/src/lib.rs]: *mut size_t

File: src/rocksdb.rs [hunk 2]
```diff
@@ -418,6 +418,39 @@ impl DB {
...
+                                                      cpath.as_ptr() as *const _,
+                                                      lencf_ptr,
+                                                      err_ptr);
+            if !err.is_null() {
+                return Err(error_message(err));
+            }
+            let list_cfs = slice::from_raw_parts(list, lencf);
+            for cf_name in list_cfs {
+                let len = libc::strlen(*cf_name);
+                let cf = String::from_utf8_lossy(slice::from_raw_parts(*cf_name as *const u8, len));
```
Comment (siddontang) [inline src/rocksdb.rs]: I think we must use valid utf8 foramt for column family name. Here we can return error.

**Comment by siddontang** [comment]
Rest LGTM

File: src/rocksdb.rs [hunk 3]
```diff
@@ -418,6 +418,39 @@ impl DB {
...
+        let mut cfs: Vec<String> = vec![];
+        unsafe {
+            let lencf: size_t = 0;
+            let lencf_ptr = &lencf as *const size_t;
+            let mut err: *const i8 = 0 as *const i8;
+            let err_ptr: *mut *const i8 = &mut err;
+            let list = rocksdb_ffi::rocksdb_list_column_families(opts.inner,
+                                                      cpath.as_ptr() as *const _,
+                                                      lencf_ptr,
+                                                      err_ptr);
```
Comment (BusyJay) [inline src/rocksdb.rs]: `&mut err` should be enough.

Comment (zhangjinpeng1987) [inline src/rocksdb.rs]: Here need type char**, if &mut err satisfied?

File: src/rocksdb.rs [hunk 4]
```diff
@@ -1288,6 +1320,32 @@ mod test {
...
+            for (&cf, &cf_opts) in cfs.iter().zip(&cfs_ref_opts) {
+                if cf == "default" {
+                    continue;
+                }
+                db.create_cf(cf, cf_opts).unwrap();
+            }
+        }
+        let opts_list_cfs = Options::new();
+        let cfs_vec = DB::list_column_families(&opts_list_cfs, path.path().to_str().unwrap()).unwrap();
+        assert!(cfs_vec.len() == 4);
```
Comment (BusyJay) [inline src/rocksdb.rs]: Use `assert_eq` instead.

File: src/rocksdb.rs [hunk 5]
```diff
@@ -418,6 +418,38 @@ impl DB {
...
+        };
+
+        let mut cfs: Vec<String> = vec![];
+        unsafe {
+            let mut lencf: size_t = 0;
+            let lencf_ptr = &mut lencf as *mut size_t;
+            let mut err: *const i8 = 0 as *const i8;
+            let list = rocksdb_ffi::rocksdb_list_column_families(opts.inner,
+                                                      cpath.as_ptr() as *const _,
+                                                      lencf_ptr,
```
Comment (BusyJay) [inline src/rocksdb.rs]: `&mut lencf` should be enough too.

File: src/rocksdb.rs [hunk 6]
```diff
@@ -1344,7 +1343,7 @@ mod test {
         }
         let opts_list_cfs = Options::new();
         let cfs_vec = DB::list_column_families(&opts_list_cfs, path.path().to_str().unwrap()).unwrap();
-        assert!(cfs_vec.len() == 4);
+        assert_eq!(cfs_vec.len(), cfs.len());
```
Comment (BusyJay) [inline src/rocksdb.rs]: Why not `assert_eq!(cfs_vec, cfs)`?

**Comment by zhangjinpeng1987** [comment]
@siddontang @ngaut PTAL

File: src/rocksdb.rs (hunk 2 above)
Comment (siddontang) [inline src/rocksdb.rs]: ping
should return error, not ignore utf8 error.

**Comment by zhangjinpeng1987** [comment]
@siddontang PTAL

File: librocksdb_sys/src/lib.rs [hunk 7]
```diff
@@ -440,6 +440,13 @@ extern "C" {
                                       column_family_handle: DBCFHandle,
                                       err: *mut *const i8);
     pub fn rocksdb_column_family_handle_destroy(column_family_handle: DBCFHandle);
+    pub fn rocksdb_list_column_families(db: DBOptions,
```
Comment (BusyJay) [inline librocksdb_sys/src/lib.rs]: Please merge master first.

File: src/rocksdb.rs [hunk 8]
```diff
@@ -440,8 +440,11 @@ impl DB {
             let list_cfs = slice::from_raw_parts(list, lencf);
             for cf_name in list_cfs {
                 let len = libc::strlen(*cf_name);
-                let cf = String::from_utf8_lossy(slice::from_raw_parts(*cf_name as *const u8, len));
-                cfs.push(format!("{}", cf));
+                let cf = match String::from_utf8(slice::from_raw_parts(*cf_name as *const u8, len).to_vec()) {
```
Comment (siddontang) [inline src/rocksdb.rs]: can we use try! directly?

Comment (zhangjinpeng1987) [inline src/rocksdb.rs]: Error type didn't match.

**Comment by siddontang** [comment]
LGTM

**Pull request merged**